pkg/raft/log.go

```diff
@@ -337,17 +337,19 @@ func (l *raftLog) lastIndex() uint64 {
 
 // commitTo bumps the commit index to the given value if it is higher than the
 // current commit index.
-//
-// TODO(pav-kv): this method should accept the term of the leader on whose
-// behalf the commit index is bumped. It is only safe to update the commit index
-// if our log is consistent with this leader, i.e. accTerm >= term.
-func (l *raftLog) commitTo(tocommit uint64) {
+func (l *raftLog) commitTo(mark logMark) {
+	// TODO(pav-kv): it is only safe to update the commit index if our log is
+	// consistent with the mark.term leader. If the mark.term leader sees the
+	// mark.index entry as committed, all future leaders have it in the log. It is
+	// thus safe to bump the commit index to min(mark.index, lastIndex) if our
```

Comment:
@nvanbenschoten You brought a good point today that we could also remember the `mark` as is. It would then automatically commit some entries later when we get appends.

Currently, we have the invariant: `l.committed <= l.lastIndex`.
But if we make `l.committed` to be a `logMark`, then it can overtake `lastIndex`. The "logical" committed index could then be computed dynamically as `max(l.committed.index, l.lastIndex())`.

Reply:
Scenario:
- A leader `t1` commits 100 entries and sends commit index = 100.
- This races with a leader `t2` being elected, and term bumps on all the followers.
- Now, it will take the new leader one RTT to commit its first entry and learn the commit index. Then another 1/2 RTT to deliver the commit index to followers. All the followers are at commit index 0 before that.

If we make `l.committed` a `logMark` as suggested, we will remove this 3/2 RTT delay from the followers learning the committed state.